src/coreComponents/fieldSpecification/FieldSpecificationManager.hpp

```diff
@@ -233,80 +222,97 @@ class FieldSpecificationManager : public dataRepository::Group
-          {
-            dataRepository::Group * const elemRegionSubGroup = targetGroup->getGroupPointer( ElementRegionManager::groupKeyStruct::elementRegionsGroup() );
-            if( elemRegionSubGroup != nullptr )
-            {
-              targetGroup = elemRegionSubGroup;
-            }
-
-            dataRepository::Group * const elemSubRegionSubGroup = targetGroup->getGroupPointer( ElementRegionBase::viewKeyStruct::elementSubRegions() );
-            if( elemSubRegionSubGroup != nullptr )
-            {
-              targetGroup = elemSubRegionSubGroup;
-            }
-
-            if( targetPath[pathLevel] == ElementRegionManager::groupKeyStruct::elementRegionsGroup() ||
-                targetPath[pathLevel] == ElementRegionBase::viewKeyStruct::elementSubRegions() )
-            {
-              continue;
-            }
-
-            targetGroup = &targetGroup->getGroup( targetPath[pathLevel] );
-          }
-          applyOnTargetRecursive( *targetGroup, fs, targetName, lambda );
-        }
+        array1d< string > const targetPath = stringutilities::tokenize( fs.getObjectPath(), "/" );
+        applyToPathRecursive( &mesh, targetPath, fs, std::forward< LAMBDA >( lambda ) );
       }
     } );
   }
 
 private:
 
   template< typename BCTYPE, typename LAMBDA >
-  void applyOnTargetRecursive( Group & target,
+  void applyToPathRecursive( dataRepository::Group * targetGroup,
+                             Span< string const > const targetPath,
+                             BCTYPE const & fs,
+                             LAMBDA && lambda ) const
+  {
+    constexpr std::array< char const *, 2 > groupsToSkip =
+    {
+      ElementRegionManager::groupKeyStruct::elementRegionsGroup(),
+      ElementRegionBase::viewKeyStruct::elementSubRegions()
+    };
+
+    for( std::size_t pathLevel = 0; pathLevel < targetPath.size(); ++pathLevel )
+    {
+      // Skip certain groups in the tree by jumping directly to them
+      for( char const * const groupName : groupsToSkip )
+      {
+        if( dataRepository::Group * const subGroup = targetGroup->getGroupPointer( groupName ) )
+        {
+          targetGroup = subGroup;
+        }
+      }
+
+      // Also skip them in the object path being processed
+      if( std::find( groupsToSkip.begin(), groupsToSkip.end(), targetPath[pathLevel] ) != groupsToSkip.end() )
+      {
+        continue;
+      }
+
+      // If the next token is a list, process it recursively and stop
+      if( targetPath[pathLevel][0] == '{' )
+      {
+        array1d< string > subGroupNames;
+        LvArray::input::stringToArray( subGroupNames, targetPath[pathLevel] );
+        for( string const & subGroupName : subGroupNames )
+        {
+          applyToPathRecursive( &targetGroup->getGroup( subGroupName ),
+                                targetPath.subspan( pathLevel + 1 ),
+                                fs, std::forward< LAMBDA >( lambda ) );
+        }
+        return;
+      }
```

Comment:
You could surely do preprocessing and consider `targetPath` as a `Span< vector< string > >`, even with a lot of arrays of length 1.
That would allow you to have one single path for your algo (and keep string parsing out of this loop)?
Plus, (it's more a matter of style and what's done in the skip is unclear to me), but I would extract the `skip` part as a lambda function outside of the `pathLevel` loop to get something really concise.

Reply:
This would be a lot of extra allocations happening though - whereas strings will mostly be SBO-optimized, vectors are not :( Not saying we should necessarily care too much (even though we pass through this code multiple times on every BC application) but would be great if this parsing only ever happened once. Ideally I'd like to store pre-parsed path as a
```c++
vector< variant< string, array1d< string > > > m_path;
```
inside the `FieldSpecification` object.

Reply:
Yes there may be a _tiny_ overhead (how much?), but also we would separate the parsing of the string from the algorithm which operates on some kind of tree; having separate responsibilities is always good. 

There is no strong reason why we should deal with "_different types of nodes_" in the tree and have a big `if` in the middle of the loop.